File: android/core/src/test/kotlin/me/raddatz/cindy/core/calibration/CalibrationAnalyzerTests.kt
package me.raddatz.cindy.core.calibration

import me.raddatz.cindy.core.Exercise
import me.raddatz.cindy.core.Sample
import me.raddatz.cindy.core.SignalConfig
import me.raddatz.cindy.core.SignalSource
import me.raddatz.cindy.core.SyntheticSignal
import me.raddatz.cindy.core.WorkoutPlan
import me.raddatz.cindy.core.signal.RepDirection
import me.raddatz.cindy.core.signal.RepThresholds
import org.junit.Test
import java.io.File
import java.nio.file.Files
import java.time.Instant
import kotlin.math.abs
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertIs
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class CalibrationAnalyzerTests {
    private fun trace(samples: List<Sample>): List<CalibrationSample> = samples.mapNotNull { sample ->
        val value = sample.value
        if (value == null || sample.confidence < 0.5f) null else CalibrationSample(sample.t, value)
    }

    private fun tempDirectory(): File = Files.createTempDirectory("cindy-calibration").toFile().apply { deleteOnExit() }

    @Test
    fun detectsOnePeakCycle() {
        val samples = SyntheticSignal(rest = 0.05f, peak = 0.20f).reps(1, period = 1.6, leadIn = 0.8, leadOut = 0.5)
        val analyzer = CalibrationAnalyzer(source = SignalSource.FACE)
        val calibration = assertNotNull(analyzer.evaluate(trace(samples)))
        assertEquals(RepDirection.PEAK, calibration.direction)
        assertTrue(abs(calibration.minValue - 0.05f) < 0.01f)
        assertTrue(abs(calibration.maxValue - 0.20f) < 0.01f)
        assertTrue(calibration.low > calibration.minValue)
        assertTrue(calibration.high < calibration.maxValue)
        assertTrue(calibration.low < calibration.high)
        assertTrue(calibration.repDuration > 0.8 && calibration.repDuration < 2.0)
        assertTrue(abs(calibration.baseline - 0.05f) < 0.01f)
    }

    @Test
    fun calibratesAtFiveFramesPerSecond() {
        // Galaxy A20e: half a second of rest holds three samples, not the five the 30 fps rule wants.
        val samples = SyntheticSignal(fps = 5.0, rest = 0.05f, peak = 0.20f).reps(1, period = 1.6, leadIn = 0.8, leadOut = 0.5)
        val calibration = assertNotNull(CalibrationAnalyzer(source = SignalSource.FACE).evaluate(trace(samples)))
        assertEquals(RepDirection.PEAK, calibration.direction)
        assertTrue(abs(calibration.baseline - 0.05f) < 0.01f)
    }

    @Test
    fun aPatchyBaselineAtThirtyFramesPerSecondIsStillRejected() {
        val samples = SyntheticSignal(rest = 0.05f, peak = 0.20f).reps(1, period = 1.6, leadIn = 0.8, leadOut = 0.5)
        // Only four confident samples in the first half second; the rest of the trace at full rate.
        val patchy = samples.mapIndexed { i, s -> if (s.t <= 0.5 && i % 4 != 0) Sample(s.t, s.value, 0f) else s }
        assertNull(CalibrationAnalyzer(source = SignalSource.FACE).evaluate(trace(patchy)))
    }

    @Test
    fun detectsTroughCycle() {
        val peak = SyntheticSignal(rest = 0.05f, peak = 0.20f).reps(1, leadIn = 0.8, leadOut = 0.5)
        val inverted = peak.map { Sample(it.t, it.value?.let { v -> 0.25f - v }, it.confidence) }
        val calibration = assertNotNull(CalibrationAnalyzer(source = SignalSource.FACE).evaluate(trace(inverted)))
        assertEquals(RepDirection.TROUGH, calibration.direction)
        assertTrue(abs(calibration.baseline - 0.20f) < 0.01f)
    }

    @Test
    fun incompleteCycleIsNotAccepted() {
        // Cut the trace right after the peak: no return to the rest band.
        val samples = SyntheticSignal().reps(1, leadIn = 0.8, leadOut = 0.0)
        var peakIndex = 0
        for (i in samples.indices) if ((samples[peakIndex].value ?: 0f) < (samples[i].value ?: 0f)) peakIndex = i
        val cut = samples.subList(0, peakIndex + 1)
        val analyzer = CalibrationAnalyzer(source = SignalSource.FACE)
        assertNull(analyzer.evaluate(trace(cut)))
        assertEquals(CalibrationFailure.NoReturn, analyzer.diagnose(trace(cut)))
    }

    @Test
    fun weakSignalIsDiagnosed() {
        val samples = SyntheticSignal(rest = 0.05f, peak = 0.055f).reps(1, leadIn = 0.8)
        val analyzer = CalibrationAnalyzer(source = SignalSource.FACE)
        assertNull(analyzer.evaluate(trace(samples)))
        assertEquals(CalibrationFailure.TooWeak, analyzer.diagnose(trace(samples)))
    }

    @Test
    fun noSamplesMeansNoFace() {
        val analyzer = CalibrationAnalyzer(source = SignalSource.FACE)
        assertEquals(CalibrationFailure.NoFace, analyzer.diagnose(emptyList()))
        assertNull(analyzer.evaluate(emptyList()))
    }

    @Test
    fun tooFastCycleIsDiagnosed() {
        val samples = SyntheticSignal().reps(1, period = 0.2, leadIn = 0.8)
        val analyzer = CalibrationAnalyzer(source = SignalSource.FACE)
        assertNull(analyzer.evaluate(trace(samples)))
        val failure = assertIs<CalibrationFailure.ImplausibleDuration>(analyzer.diagnose(trace(samples)))
        assertTrue(failure.duration < 0.5)
    }

    @Test
    fun profileRoundTripsThroughJSON() {
        val store = CalibrationStore(File(tempDirectory(), "calibration.json"))
        var profile = CalibrationProfile()
        assertFalse(profile.isComplete)
        for (exercise in Exercise.entries) {
            profile = profile.withCalibration(
                ExerciseCalibration(
                    source = SignalConfig.default.source(exercise), minValue = 0.1f, maxValue = 0.3f, baseline = 0.1f,
                    low = 0.15f, high = 0.25f, direction = exercise.defaultRepDirection, repDuration = 1.2,
                    calibratedAt = Instant.now(),
                ),
                exercise,
            )
        }
        assertTrue(profile.isComplete)
        store.save(profile)
        val loaded = assertNotNull(store.load())
        assertTrue(loaded.isComplete)
        assertEquals(RepDirection.TROUGH, loaded.calibration(Exercise.PULL_UP)?.direction)
        assertEquals(
            RepThresholds(0.15f, 0.25f, RepDirection.PEAK, baseline = 0.1f),
            loaded.calibration(Exercise.SQUAT)?.thresholds,
        )
    }

    @Test
    fun calibrationsOnAnOldSignalSourceAreDropped() {
        val store = CalibrationStore(File(tempDirectory(), "calibration.json"))
        var profile = CalibrationProfile()
        for (exercise in Exercise.entries) {
            // Every exercise on the face, as squats were calibrated before they moved to body pose.
            profile = profile.withCalibration(
                ExerciseCalibration(
                    source = SignalSource.FACE, minValue = 0.1f, maxValue = 0.3f, baseline = 0.1f, low = 0.15f,
                    high = 0.25f, direction = exercise.defaultRepDirection, repDuration = 1.2,
                    calibratedAt = Instant.now(),
                ),
                exercise,
            )
        }
        store.save(profile)
        val loaded = assertNotNull(store.load())
        assertEquals(listOf(Exercise.SQUAT), loaded.missingExercises(WorkoutPlan.cindy))
        assertNotNull(loaded.calibration(Exercise.PUSH_UP))
    }

    @Test
    fun thePlankNeedsNoCalibration() {
        val plan = WorkoutPlan.cindy.withEnabled(Exercise.PLANK, true)
        var profile = CalibrationProfile()
        for (exercise in WorkoutPlan.cindy.exercises) {
            profile = profile.withCalibration(
                ExerciseCalibration(
                    source = SignalConfig.default.source(exercise), minValue = 0.1f, maxValue = 0.3f, baseline = 0.1f,
                    low = 0.15f, high = 0.25f, direction = exercise.defaultRepDirection, repDuration = 1.2,
                    calibratedAt = Instant.now(),
                ),
                exercise,
            )
        }
        assertTrue(profile.missingExercises(plan).isEmpty())
        assertTrue(profile.isComplete(plan))
        // A plank calibration from before the timer is dropped on load.
        profile = profile.withCalibration(
            ExerciseCalibration(
                source = SignalConfig.default.source(Exercise.PLANK), minValue = 0.1f, maxValue = 0.3f,
                baseline = 0.2f, low = 0.15f, high = 0.25f, direction = RepDirection.PEAK, repDuration = 3.0,
                calibratedAt = Instant.now(),
            ),
            Exercise.PLANK,
        )
        assertNull(profile.removingOutdated(SignalConfig.default).calibration(Exercise.PLANK))
    }
}

File: ios/CindyTests/CalibrationAnalyzerTests.swift
import Foundation
import Testing
@testable import Cindy

struct CalibrationAnalyzerTests {
    private func trace(_ samples: [SyntheticSignal.Sample]) -> [CalibrationSample] {
        samples.compactMap { sample in
            guard let value = sample.value, sample.confidence >= 0.5 else { return nil }
            return CalibrationSample(timestamp: sample.t, value: value)
        }
    }

    @Test func detectsOnePeakCycle() throws {
        let samples = SyntheticSignal(rest: 0.05, peak: 0.20).reps(1, period: 1.6, leadIn: 0.8, leadOut: 0.5)
        let analyzer = CalibrationAnalyzer(source: .face)
        let calibration = try #require(analyzer.evaluate(trace(samples)))
        #expect(calibration.direction == .peak)
        #expect(abs(calibration.minValue - 0.05) < 0.01)
        #expect(abs(calibration.maxValue - 0.20) < 0.01)
        #expect(calibration.low > calibration.minValue)
        #expect(calibration.high < calibration.maxValue)
        #expect(calibration.low < calibration.high)
        #expect(calibration.repDuration > 0.8 && calibration.repDuration < 2.0)
        #expect(abs(calibration.baseline - 0.05) < 0.01)
    }

    @Test func calibratesAtFiveFramesPerSecond() throws {
        // Galaxy A20e: half a second of rest holds three samples, not the five the 30 fps rule wants.
        let samples = SyntheticSignal(fps: 5, rest: 0.05, peak: 0.20).reps(1, period: 1.6, leadIn: 0.8, leadOut: 0.5)
        let calibration = try #require(CalibrationAnalyzer(source: .face).evaluate(trace(samples)))
        #expect(calibration.direction == .peak)
        #expect(abs(calibration.baseline - 0.05) < 0.01)
    }

    @Test func aPatchyBaselineAtThirtyFramesPerSecondIsStillRejected() {
        let samples = SyntheticSignal(rest: 0.05, peak: 0.20).reps(1, period: 1.6, leadIn: 0.8, leadOut: 0.5)
        // Only four confident samples in the first half second; the rest of the trace at full rate.
        let patchy = samples.enumerated().map { i, s in
            s.t <= 0.5 && i % 4 != 0 ? SyntheticSignal.Sample(t: s.t, value: s.value, confidence: 0) : s
        }
        #expect(CalibrationAnalyzer(source: .face).evaluate(trace(patchy)) == nil)
    }

    @Test func detectsTroughCycle() throws {
        let peak = SyntheticSignal(rest: 0.05, peak: 0.20).reps(1, leadIn: 0.8, leadOut: 0.5)
        let inverted = peak.map { SyntheticSignal.Sample(t: $0.t, value: $0.value.map { 0.25 - $0 }, confidence: $0.confidence) }
        let calibration = try #require(CalibrationAnalyzer(source: .face).evaluate(trace(inverted)))
        #expect(calibration.direction == .trough)
        #expect(abs(calibration.baseline - 0.20) < 0.01)
    }

    @Test func incompleteCycleIsNotAccepted() {
        // Cut the trace right after the peak: no return to the rest band.
        let samples = SyntheticSignal().reps(1, leadIn: 0.8, leadOut: 0)
        let peakIndex = samples.indices.max { (samples[$0].value ?? 0) < (samples[$1].value ?? 0) }!
        let cut = Array(samples[...peakIndex])
        let analyzer = CalibrationAnalyzer(source: .face)
        #expect(analyzer.evaluate(trace(cut)) == nil)
        #expect(analyzer.diagnose(trace(cut)) == .noReturn)
    }

    @Test func weakSignalIsDiagnosed() {
        let samples = SyntheticSignal(rest: 0.05, peak: 0.055).reps(1, leadIn: 0.8)
        let analyzer = CalibrationAnalyzer(source: .face)
        #expect(analyzer.evaluate(trace(samples)) == nil)
        #expect(analyzer.diagnose(trace(samples)) == .tooWeak)
    }

    @Test func noSamplesMeansNoFace() {
        let analyzer = CalibrationAnalyzer(source: .face)
        #expect(analyzer.diagnose([]) == .noFace)
        #expect(analyzer.evaluate([]) == nil)
    }

    @Test func tooFastCycleIsDiagnosed() {
        let samples = SyntheticSignal().reps(1, period: 0.2, leadIn: 0.8)
        let analyzer = CalibrationAnalyzer(source: .face)
        #expect(analyzer.evaluate(trace(samples)) == nil)
        if case .implausibleDuration(let duration) = analyzer.diagnose(trace(samples)) {
            #expect(duration < 0.5)
        } else {
            Issue.record("expected implausibleDuration")
        }
    }

    @Test func profileRoundTripsThroughJSON() throws {
        let directory = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString)
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        let store = CalibrationStore(url: directory.appendingPathComponent("calibration.json"))
        var profile = CalibrationProfile()
        #expect(profile.isComplete == false)
        for exercise in Exercise.allCases {
            profile.set(ExerciseCalibration(source: SignalConfig.default.source(for: exercise), minValue: 0.1, maxValue: 0.3, baseline: 0.1, low: 0.15,
                                            high: 0.25, direction: exercise.defaultRepDirection, repDuration: 1.2,
                                            calibratedAt: Date()), for: exercise)
        }
        #expect(profile.isComplete)
        try store.save(profile)
        let loaded = try #require(store.load())
        #expect(loaded.isComplete)
        #expect(loaded.calibration(for: .pullUp)?.direction == .trough)
        #expect(loaded.calibration(for: .squat)?.thresholds == RepThresholds(low: 0.15, high: 0.25, direction: .peak,
                                                                                baseline: 0.1))
    }

    @Test func calibrationsOnAnOldSignalSourceAreDropped() throws {
        let directory = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString)
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        let store = CalibrationStore(url: directory.appendingPathComponent("calibration.json"))
        var profile = CalibrationProfile()
        for exercise in Exercise.allCases {
            // Every exercise on the face, as squats were calibrated before they moved to body pose.
            profile.set(ExerciseCalibration(source: .face, minValue: 0.1, maxValue: 0.3, baseline: 0.1, low: 0.15,
                                            high: 0.25, direction: exercise.defaultRepDirection, repDuration: 1.2,
                                            calibratedAt: Date()), for: exercise)
        }
        try store.save(profile)
        let loaded = try #require(store.load())
        #expect(loaded.missingExercises(for: .cindy) == [.squat])
        #expect(loaded.calibration(for: .pushUp) != nil)
    }

    @Test func thePlankNeedsNoCalibration() {
        var plan = WorkoutPlan.cindy
        plan.setEnabled(.plank, true)
        var profile = CalibrationProfile()
        for exercise in WorkoutPlan.cindy.exercises {
            profile.set(ExerciseCalibration(source: SignalConfig.default.source(for: exercise), minValue: 0.1,
                                            maxValue: 0.3, baseline: 0.1, low: 0.15, high: 0.25,
                                            direction: exercise.defaultRepDirection, repDuration: 1.2,
                                            calibratedAt: Date()), for: exercise)
        }
        #expect(profile.missingExercises(for: plan).isEmpty)
        #expect(profile.isComplete(for: plan))
        // A plank calibration from before the timer is dropped on load.
        profile.set(ExerciseCalibration(source: SignalConfig.default.source(for: .plank), minValue: 0.1, maxValue: 0.3,
                                        baseline: 0.2, low: 0.15, high: 0.25, direction: .peak, repDuration: 3,
                                        calibratedAt: Date()), for: .plank)
        #expect(profile.removingOutdated(for: .default).calibration(for: .plank) == nil)
    }
}
